File: Homework/Ngan_Luu/841-keys-and-rooms.py
# 841. https://leetcode.com/problems/keys-and-rooms/
class Solution:                
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited=[None]*len(rooms)
        def dfs(curr):
            visited[curr]=True
            print(curr)
            for i in rooms[curr]:
                if not(visited[i]):
                    dfs(i)
        dfs(0)
        return all(visited)

File: Homework/Son/W3/841-keys-and-rooms.js
var canVisitAllRooms = function (g) {
    let V = g.length;

    let visitedAll = new Array(V).fill(false);
    let visitedCount = 0;

    const dfs = (u) => {
        if (!visitedAll[u]) {
            visitedAll[u] = true;

            visitedCount++;
            for (let v of g[u]) {
                dfs(v);
            }
        }
    };

    dfs(0);
    return visitedCount == V;
};

// https://leetcode.com/problems/keys-and-rooms/